doc/man/mk.1
.\" Generated by mkdoc on January, 2018
.TH "MK" "1" "January, 2018" "mk 1.3.12" "User Commands"
.de nl
.sp 0
..
.de hr
.sp 1
.nf
.ce
.in 4
\l’80’
.fi
..
.de h1
.RE
.sp 1
\fB\\$1\fR
.RS 4
..
.de h2
.RE
.sp 1
.in 4
\fB\\$1\fR
.RS 6
..
.de h3
.RE
.sp 1
.in 6
\fB\\$1\fR
.RS 8
..
.de h4
.RE
.sp 1
.in 8
\fB\\$1\fR
.RS 10
..
.de h5
.RE
.sp 1
.in 10
\fB\\$1\fR
.RS 12
..
.de h6
.RE
.sp 1
.in 12
\fB\\$1\fR
.RS 14
..
.h1 "NAME"
.P
mk \- task runner
.nl
.h1 "SYNOPSIS"
.P
mk [\-h] [\-\-tasks] [\-\-help] [\-\-version] [\-\-file=<file...>] [task...]
.nl
.h1 "DESCRIPTION"
.P
Runs tasks defined in a task file. Searches for \fBmkdoc.js\fR in the current directory (or parent directories) and executes the specified tasks.
.nl
.h1 "OPTIONS"
.TP
\fB\-f, \-\-file\fR=[\fIFILE\fR...]
 Load specific task files
.nl
.TP
\fB\-\-tasks\fR
 Print task list
.nl
.TP
\fB\-h, \-\-help\fR
 Display help and exit
.nl
.TP
\fB\-\-version\fR
 Print the version and exit
.nl
.h1 "ENVIRONMENT"
.P
If TASK_FILE is set it changes the name of the file searched for, default is \fBmkdoc.js\fR.
.nl
.h1 "CREATING TASKS"
.P
Tasks are \fInamed\fR functions that are passed to the \fBtask\fR function:
.nl
.PP
.in 12
var mk = require('mktask');
.br

.br
// @task readme build the readme file.
.br
function readme(cb) {
.br
  // implement task logic
.br
  cb();
.br
}
.br

.br
mk.task(readme);
.P
Anonymous functions are not allowed and will generate an error if used.
.nl
.h1 "TASK DOCUMENTATION"
.P
It is considered good practice to annotate your tasks with comments that provide a name and description of the task so that it will be included in the list printed when running \fBmk \-\-tasks\fR.
.nl
.PP
.in 12
// @task readme build the readme file.
.h1 "TASK NAMES"
.P
By default the task identifier (referenced on the command line) is taken from the function name but you may explicitly specify an identifier if you prefer:
.nl
.PP
.in 12
mk.task('docs', function readme(cb){cb()});
.P
If you have dependencies the identifier comes afterwards:
.nl
.PP
.in 12
mk.task([api, example], 'docs', function readme(cb){cb()});
.P
When multiple tasks are passed then the identifier is taken from the last function which in this case becomes \fBreadme\fR:
.nl
.PP
.in 12
mk.task(function api(cb){cb()}, function readme(cb){cb()});
.h1 "MAIN TASK"
.P
The \fBmk\fR program when executed with no arguments will either run all available tasks in series or a \fBmain\fR task if declared. To declare a main task give it the name \fBmain\fR:
.nl
.PP
.in 12
var mk = require('mktask');
.br

.br
// @task main build all documentation.
.br
function main(cb) {
.br
  // implement task logic
.br
  cb();
.br
}
.br

.br
mk.task(main);
.h1 "DEFERRED TASKS"
.P
Typically task functions will invoke the callback function when done but they may also return an array of task functions which is useful when a task wishes to defer to a series of other tasks:
.nl
.PP
.in 12
var mk = require('mktask');
.br

.br
// @task api build the api docs.
.br
function api(cb) {
.br
  // implement api task logic
.br
  cb();
.br
}
.br

.br
// @task readme build the readme file.
.br
function readme(cb) {
.br
  // implement readme task logic
.br
  cb();
.br
}
.br

.br
// @task main build the api and readme docs.
.br
function main() {
.br
  return [api, readme];
.br
}
.br

.br
mk.task(api);
.br
mk.task(readme);
.br
mk.task(main);
.P
Note that when deferring to other task functions they must have been registered by calling \fBtask()\fR.
.nl
.h1 "STREAM TASKS"
.P
Sometimes when creating complex stream pipelines it is useful to return streams so that parts of the pipeline become reusable between tasks, for example:
.nl
.PP
.in 12
var mk = require('mktask')
.br
  , ast = require('mkast');
.br

.br
function in() {
.br
  return mk
.br
    .src('This is a markdown paragraph.')
.br
    .pipe(ast.stringify());
.br
}
.br

.br
function out() {
.br
  return mk.dest('target/stream\-example.md');
.br
}
.br

.br
mk.task(in, out);
.P
When a task returns a stream it is piped to the next task function in the pipeline and the callback function is added as a listener for the \fBfinish\fR event on the last stream in the pipeline.
.nl
.h1 "TASK DEPENDENCIES"
.P
Task functions may declare an array of functions to call before the task function(s).
.nl
.P
Dependencies are executed in parallel but they must all complete before the tasks are executed:
.nl
.PP
.in 12
var mk = require('mktask');
.br

.br
// @task api build the api docs.
.br
function api(
.br
  // implement api task logic
.br
  cb();
.br
}
.br

.br
// @task example build the example file.
.br
function example(
.br
  // implement example task logic
.br
  cb();
.br
}
.br

.br
// @task readme build the readme file.
.br
function readme(cb) {
.br
  // implement readme task logic
.br
  cb();
.br
}
.br

.br
mk.task([api, example], readme);
.h1 "TASK ARGUMENTS"
.P
Task functions are automatically exposed the parsed arguments object via \fBthis.args\fR such that \fBmk readme \-\-env devel\fR would result in the readme task being able to access the \fBenv\fR option using \fBthis.args.options.env\fR.
.nl
.P
Flags are available in \fBthis.args.flags\fR such that \fBmk readme \-v\fR yields \fBtrue\fR for \fBthis.args.flags.v\fR.
.nl
.P
Note that some command line arguments are handled by the \fBmk\fR program you should take care that the names do not conflict.
.nl
.P
For detailed information on the \fBargs\fR object see the [argparse library][argparse].
.nl
.h1 "EXAMPLE"
.P
Build all tasks (or a main task when defined):
.nl
.PP
.in 12
mk
.P
Build specific tasks:
.nl
.PP
.in 12
mk api readme
.P
Use a specific build file:
.nl
.PP
.in 12
mk \-f ~/mkdoc.js readme
.P
To see a list of tasks use:
.nl
.PP
.in 12
mk \-\-tasks

cli/mk.js
var fs = require('fs')
  , path = require('path')
  , vm = require('vm')
  , EOL = require('os').EOL
  , mk = require('mktask')
  , mkparse = require('mkparse')
  , NAME = process.env.TASK_FILE || 'mkdoc.js'
  , cli = require('mkcli-runtime')
  , pkg = require('mktask/package.json')
  , prg = cli.load(require('../doc/json/mk.json'));

/**
 *  Execute an external command synchronously or asynchronously
 *  when a callback is given.
 */
function exec(cmd, options, cb) {
  var ps = require('child_process')
    , ex = (typeof cb === 'function') ? ps.exec : ps.execSync;
  return ex(cmd, options, cb);
}

function command(program) {
  return function(cmd, options, cb) {
    if(program) {
      cmd = program + ' ' + cmd;
    }
    return exec(cmd, options, cb);
  }
}

function print(files, runner, output, cb) {
  var list = files.slice();

  function next(err) {

    /* istanbul ignore next: files are validated earlier so tough to mock */
    if(err) {
      return cb(err);
    }

    var file = list.shift()
      , tags = [];
    if(!file) {
      return cb();
    }

    var parse = mkparse.load(file, function complete() {
      var max = tags.reduce(function(prev, item) {
        return Math.max(prev, item.name.length);
      }, 0);
      tags.forEach(function(tag) {
        var name = tag.name;
        while(name.length < max) {
          name += ' ';
        }
        output.write(name + '  ' + tag.description);
        if(tag.missing) {
          output.write(' (missing)');
        }
        output.write(EOL);
      })
      next();
    });
    parse.on('comment', function(comment) {
      var missing;
      //console.dir(comment);
      comment.tags.forEach(function(tag) {
        missing = false;
        if(tag.id === 'task') {
          if(!runner.get(tag.name)) {
            missing = true;
          }
          tags.push(
            {name: tag.name, description: tag.description, missing: missing});
        }
      })
    })
  }

  next();
}

/**
 *  @name mk
 *  @cli doc/cli/mk.md
 */
function main(argv, conf, cb) {

  /* istanbul ignore if: always pass argv in test env */
  if(typeof argv === 'function') {
    cb = argv;
    argv = null;
  }

  /* istanbul ignore if: always pass conf in test env */
  if(typeof conf === 'function') {
    cb = conf;
    conf = null;
  }

  /* istanbul ignore next: always pass conf in test env */
  conf = conf || {};
  /* istanbul ignore next: never print to stdout in test env */
  conf.output = conf.output || process.stdout;
  /* istanbul ignore next: never read from stdin in test env */
  conf.input = conf.input || process.stdin;

  var opts = {
      input: conf.input,
      output: conf.output
    }
    , runtime = {
        base: path.normalize(path.join(__dirname, '..')),
        target: opts,
        hints: prg,
        help: {
          file: 'doc/help/mk.txt',
          output: conf.output
        },
        version: {
          name: pkg.name.replace('task', ''),
          version: pkg.version,
          output: conf.output
        },
        plugins: [
          require('mkcli-runtime/plugin/epipe'),
          require('mkcli-runtime/plugin/hints'),
          require('mkcli-runtime/plugin/argv'),
          require('mkcli-runtime/plugin/help'),
          require('mkcli-runtime/plugin/version')
        ]
      };

  cli.run(prg, argv, runtime, function parsed(err, req) {
    if(err || req.aborted) {
      return cb(err);
    }

    var deps = {
      abs: require('mkabs'),
      api: require('mkapi'),
      ast: require('mkast'),
      cat: require('mkcat'),
      cli: require('mkcli'),
      exe: require('../lib/exe'),
      filter: require('mkfilter'),
      level: require('mklevel'),
      msg: require('mkmsg'),
      out: require('mkout'),
      page: require('mkpage'),
      parse: require('mkparse'),
      pi: require('mkpi'),
      ql: require('mkql'),
      ref: require('mkref'),
      toc: require('mktoc')
    };

    // utility to execute programs
    mk.exec = exec;

    // utility to bind a command to a program
    mk.command = command;

    // decorate with `doc` function
    mk.doc = require('../index');

    // static stream creation access
    for(var k in deps) {
      mk[k] = deps[k];
    }

    var dir = process.cwd()
      , file = path.join(dir, NAME)
      , files = this.file
      , tasks
      , list
      , called = false
      , stat;

    function done(err) {
      if(!called) {
        called = true;
        cb(err);
      }
    }

    function get(file, strict) {
      try {
        stat = fs.statSync(file)
        // ok if the file does not exist, searching parents
      }catch(e) {}

      if(strict && (!stat || stat && !stat.isFile())) {
        return done(new Error('invalid file: ' + file));
      }

      function req(mod) {

        // return our reference to `mktask` so there are not
        // module conflicts
        if(mod === 'mktask') {
          return mk;
        }

        // resolve relative paths
        if(!path.isAbsolute(mod) && /^\.\.?\//.test(mod)) {
          mod = path.join(path.dirname(file), mod);
        }

        // console.error('require proxy %s', mod);

        try {
          return require(mod);
        }catch(e) {
          //console.error(e);
         // console.dir('trying relative path...');
          // try path relative to build file
          mod = path.join(path.dirname(file), 'node_modules', mod);
          return require(mod);
        }
      }

      var context = vm.createContext(
          {
            __filename: file,
            __dirname: path.dirname(file),
            Buffer: Buffer,
            require: req,
            console: console,
            process: process,
            setTimeout: setTimeout,
            setInterval: setInterval,
            clearTimeout: clearTimeout,
            clearInterval: clearInterval,
            setImmediate: setImmediate,
            clearImmediate: clearImmediate
          });

      if(stat && stat.isFile()) {
        try {
          var contents = fs.readFileSync(file).toString('utf8')
            , script = new vm.Script(contents);
          script.runInContext(context, {filename: file});
          return true;
        }catch(e) {
          return done(e);
        }
      }
    }

    if(files) {

      if(!Array.isArray(files)) {
        files = [files];
      }

      files.forEach(function(file) {
        if(!/^\.?\//.test(file)) {
          file = path.join(process.cwd(), file);
        }
        tasks = get(file, true);
      })

    }else{
      while(!(tasks = get(file))) {
        dir = path.dirname(dir);
        file = path.join(dir, NAME);
        if(dir === '/') {
          break;
        }
      }

      // change to parent directory of mkdoc.js
      // so that input and output paths are relative to
      // the mkdoc.js file and not the cwd when in a deeper
      // directory
      if(tasks) {
        process.chdir(path.dirname(file));
      }
    }

    if(!tasks) {
      return done(
        new Error(
          'no task file (' + NAME + ') found in ' + process.cwd()
          + ' or any parent directories '
        ));
    }

    var collection = mk.task()
      , runner;

    if(!collection.tasks.length) {
      return done(
        new Error(
          'task file ' + file + ' does not define task functions'
        ));
    }

    runner = collection.run();
    list = req.unparsed;

    if(this.tasks) {

      // print from auto-detected file
      if(!files) {
        files = [file];
      }

      return print(files, runner, conf.output, done);
    }

    for(var i = 0;i < list.length;i++) {
      if(!runner.get(list[i])) {
        return done(
          new Error('task not found: ' + list[i]));
      }
    }

    // set up execution scope for default task collection
    runner.scope = {args: req.args};

    runner.each(list, done);
  })
}

main.pkg = pkg;

module.exports = main;
